+neurostim/plugin.m

```diff
@@ -807,14 +805,109 @@ function guiSet(o,parms) %#ok<INUSD>
+            m= metaclass(current);
+            dependent = [m.PropertyList.Dependent];
+            % Find properties that we can set now (based on the stored fromFile object)            
+            storedFn = fieldnames(old);
+            missingInSaved  = strcat(setdiff({m.PropertyList(~dependent).Name},storedFn),' / ');
+            missingInCurrent  = strcat(setdiff(storedFn,{m.PropertyList(~dependent).Name}),' / ');
+            toCopy= intersect(storedFn,{m.PropertyList(~dependent).Name});
+            fprintf('Fixing backward compatibility of stored ***%s***object.\n',m.Name)
+            if ~isempty(missingInSaved)
+                fprintf('Not defined when saved (will get current default values):\n ');
+                fprintf('\t%s',missingInSaved{:})
+                fprintf('\n')
+            end
+            if ~isempty(missingInCurrent)
+                fprintf('Not defined currently (will be removed):\n');
+                fprintf('\t%s' ,missingInCurrent{:})
+                fprintf('\n');
+            end
+            for i=1:numel(toCopy)
+                try
+                    current.(toCopy{i}) = old.(toCopy{i});
+                catch me
+                    fprintf('\t Failed to set %s(will get current default value): %s\n', toCopy{i},me.message)
+                end
+            end
+            
+            
+            % Restore dynamic properties which appear to be lost (probably
+            % because a struct (old) cannot have dynprops            
+            prmsNames= fieldnames(current.prms);
+            for p=1:numel(prmsNames)
+                hDynProp = findprop(current,prmsNames{p});
+                if isempty(hDynProp)
+                    % Stored object had a dynprop that is no longer
+                    % constructed now. Recreate it (ignoring anys special
+                    % options)
+                    hDynProp = addprop(current,prmsNames{p});
+                end
+                % Delete the handle to the dynprop that was stored (but not
+                % restored on load)
+                delete(current.prms.(prmsNames{p}).hDynProp)
+                % Then link the neurostim.parameter to the dynprop created
+                % in the default constructor call using the current
+                % classdef (current) (By now current.prms has the values
+                % corresponding to the saved object)
+                current.prms.(prmsNames{p}).hDynProp = hDynProp;
+                % And set the dynprop to return only the current (i.e.
+                % last) value in the neurostim.parameter and never update.
+                hDynProp.GetMethod = @(varargin) (current.prms.(prmsNames{p}).value);
+                hDynProp.SetMethod = @(varargin) (NaN);
+                
+                % Then make sure the .plg member of the parameter links to
+                % the newly updated plg
+                if p==1
+                    % This still points to the old style object. It will be

```

Comment:
@bartkrekelberg I'm working my way through this... why does it only do this step when p==1 (i.e. for the first parameter object)?

Reply:
Because each of the parameter objects in the plugin reference the same (old) plugin. We only have to delete the plugin once (and this does it when looping over the first parameter p==1)